sdk/storage/Azure.Storage.Common/src/AsyncCollection.cs

```diff
@@ -0,0 +1,207 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for
+// license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Storage;
+
+// TODO: Move these types into the Azure.Core package next release?
+
+namespace Azure
+{
+    /// <summary>
+    /// A collection of values that may take multiple service requests to
+    /// iterate over.
+    /// </summary>
+    /// <typeparam name="T">The type of the values.</typeparam>
+    public abstract class AsyncCollection<T> : IAsyncEnumerable<Response<T>>, IEnumerable<Response<T>>
```

Comment:
I wonder if we need IEnumerable<Response<T>> on this type at all. I can be implemented on `StorageAsyncCollection`

Reply:
I think we want this on the main public type.   Even if it's implemented explicitly, customers should be able to see it's supported.

Reply:
It's an easy way to slip into sync codepath even after calling an async client method.

Reply:
That's a fair point - `Enumerable` extension methods are going to make this pretty easy for people to mess up.  What are your thoughts @KrzysztofCwalina?

Reply:
Changing to hide the implementation of `IEnumerable<T>` which is a better bet until we decide on the Sync half of the story, as discussed.